### PR TITLE
feat: enable window pushdowns

### PR DESCRIPTION
**Pull request by jlapacik**

### CODE REVIEW

File: query/stdlib/influxdata/influxdb/rules.go [hunk 1]
```diff
@@ -31,7 +31,7 @@ func init() {
 		PushDownWindowAggregateByTimeRule{},
 		PushDownBareAggregateRule{},
 		GroupWindowAggregateTransposeRule{},
-		PushDownGroupAggregateRule{},
+		// PushDownGroupAggregateRule{},
```
Comment (jlapacik) [inline query/stdlib/influxdata/influxdb/rules.go:34]: This PR does not include grouped aggregates. This rule still needs to be enabled and the corresponding launcher tests unskipped.

File: tsdb/engine/tsm1/array_cursor.gen.go.tmpl [hunk 2]
```diff
@@ -192,6 +192,8 @@ func (c *{{$type}}) reset(seek, end int64, cacheValues Values, tsmKeyCursor *Key
 		})
 		if c.cache.pos == len(c.cache.values) {
 			c.cache.pos--
+        } else if c.cache.values[c.cache.pos].UnixNano() == seek {
+            c.cache.pos--
```
Comment (jlapacik) [inline tsdb/engine/tsm1/array_cursor.gen.go.tmpl:196]: DescendingArrayCursors need to exclude points that fall on the right side of the range boundary.

File: tsdb/engine/tsm1/array_cursor.gen.go.tmpl [hunk 3]
```diff
@@ -288,9 +292,9 @@ func (c *{{$type}}) Next() {{$arrayType}} {
 		}
 	}
 
-	if pos > 0 && c.res.Timestamps[pos-1] <= c.end {
+	if pos > 0 && c.res.Timestamps[pos-1] < c.end {
 		pos -= 2
-		for pos >= 0 && c.res.Timestamps[pos] <= c.end {
+		for pos >= 0 && c.res.Timestamps[pos] < c.end {
```
Comment (jlapacik) [inline tsdb/engine/tsm1/array_cursor.gen.go.tmpl:297]: DescendingArrayCursors need to include points that fall on the left side of the range boundary.

File: go.mod [hunk 4]
```diff
@@ -47,7 +47,7 @@ require (
 	github.com/hashicorp/vault/api v1.0.2
 	github.com/imdario/mergo v0.3.9 // indirect
 	github.com/influxdata/cron v0.0.0-20191203200038-ded12750aac6
-	github.com/influxdata/flux v0.91.0
+	github.com/influxdata/flux v0.91.1-0.20201029180714-1e9bfd49f21c
```
Comment (wolffcm) [inline go.mod:50]: I thought that we are only upgrading Flux in OSS to actual releases? @rockstar Can you confirm this?

File: query/stdlib/influxdata/influxdb/rules.go (hunk 1 above)
Comment (wolffcm) [reply query/stdlib/influxdata/influxdb/rules.go:34]: Ok. By the way `first` and `last` were not supposed to be a part of this issue either. There were separate issues for them. I'm not complaining though, it's nice to get them out of the way.

File: storage/flux/reader.go [hunk 5]
```diff
@@ -231,276 +230,6 @@ READ:
...
-// isSelector returns true if given a procedure kind that represents a selector operator.
-func isSelector(kind plan.ProcedureKind) bool {
-	return kind == FirstKind || kind == LastKind || kind == MinKind || kind == MaxKind
-}
-
-func isAggregateCount(kind plan.ProcedureKind) bool {
-	return kind == CountKind
-}
-
-func (wai *windowAggregateIterator) handleRead(f func(flux.Table) error, rs storage.ResultSet) error {
```
Comment (wolffcm) [inline storage/flux/reader.go:331]: Why is all this code being deleted? I would not have expected that. Maybe it will become clear as I continue my review.

File: storage/flux/reader.go [hunk 6]
```diff
@@ -593,19 +328,19 @@ READ:
 		done := make(chan struct{})
 		switch typedCur := cur.(type) {
 		case cursors.IntegerArrayCursor:
-			cols, defs := determineTableColsForGroup(gc.Keys(), flux.TInt)
+			cols, defs := determineTableColsForGroup(gc.Keys(), flux.TInt, gc.Aggregate(), key)
```
Comment (wolffcm) [inline storage/flux/reader.go:331]: `gc.Aggregate()` should always be `nil` here, correct?

File: storage/flux/reader.go [hunk 7]
```diff
@@ -819,6 +601,234 @@ func groupKeyForGroup(kv [][]byte, spec *query.ReadGroupSpec, bnds execute.Bound
 	return execute.NewGroupKey(cols, vs)
 }
 
+type windowAggregateIterator struct {
```
Comment (wolffcm) [inline storage/flux/reader.go:604]: Okay, so you moved all of this code? That makes it difficult to see what changed. Did you have to change a lot?

File: storage/flux/table.gen.go.tmpl [hunk 8]
```diff
@@ -360,29 +703,211 @@ func (t *{{.name}}GroupTable) Do(f func(flux.ColReader) error) error {
...
-	t.appendTags(cr)
-	t.appendBounds(cr)
+	ts, v := aggregate(arr.Timestamps, arr.Values)
+	timestamps, values := []int64{ts}, []{{.Type}}{v}
+	for {
+		arr = t.cur.Next()
+		if arr.Len() > 0 {
+			ts, v := aggregate(arr.Timestamps, arr.Values)
+			timestamps = append(timestamps, ts)
+			values = append(values, v)
```
Comment (wolffcm) [inline storage/flux/table.gen.go.tmpl:754]: Perhaps not relevant for this PR, but there is a bug here for the `ReadGroup(min/max)` case because the tag keys for the series (pre-grouping) are not preserved for selectors like min and max. @adrian-thurston is currently working on a fix for the 2.0 Cloud, and we'll need to ensure that fix also lands here once group agg pushdowns are enabled.

File: tsdb/engine/tsm1/array_cursor.gen.go.tmpl (hunk 2 above)
Comment (wolffcm) [reply tsdb/engine/tsm1/array_cursor.gen.go.tmpl:196]: This is a port of a bug fix that you already fixed elsewhere, correct? Do we have a regression test here on this branch?

Comment (wolffcm) [reply tsdb/engine/tsm1/array_cursor.gen.go.tmpl:196]: Ah, I see lots of tests below.

File: tsdb/engine/tsm1/array_cursor.gen.go.tmpl (hunk 3 above)
Comment (wolffcm) [reply tsdb/engine/tsm1/array_cursor.gen.go.tmpl:297]: I have the same question here as above, this is a port of a bug fix from elsewhere, right?

**Review by wolffcm — COMMENTED** [review]
Nothing stands out to me as incorrect here, but this is a very large set of changes, so I have a few questions before you merge this:

How much of this was copy/pasted and how much needed to be edited by hand to get it to work right?

How much of these changes are covered by tests that are on this branch?

Is there any part of this PR that you'd like me to look more closely at?

File: storage/flux/table.gen.go.tmpl (hunk 8 above)
Comment (wolffcm) [reply storage/flux/table.gen.go.tmpl:754]: I guess I'd like to be sure that we fix the `ReadGroup(min/max)` bug *before* we turn on pushdowns that expose the bug. How can we use our process to ensure this, any thoughts?

File: tsdb/engine/tsm1/array_cursor.gen.go.tmpl (hunk 2 above)
Comment (jlapacik) [reply tsdb/engine/tsm1/array_cursor.gen.go.tmpl:196]: Yes we had a launcher test that failed when it was unskipped. That test now passes and I brought over the specific unit tests for it as well.

File: tsdb/engine/tsm1/array_cursor.gen.go.tmpl (hunk 3 above)
Comment (jlapacik) [reply tsdb/engine/tsm1/array_cursor.gen.go.tmpl:297]: Correct.

File: go.mod (hunk 4 above)
Comment (jlapacik) [reply go.mod:50]: There was a bug in one of the flux window functions that the aggregate array cursors call. A few of the launcher test caught it. These changes depend on that fix.

File: storage/flux/reader.go (hunk 6 above)
Comment (jlapacik) [reply storage/flux/reader.go:331]: Because we're not pushing down grouped aggregates at the moment, yes.

File: storage/flux/reader.go (hunk 7 above)
Comment (jlapacik) [reply storage/flux/reader.go:604]: My bad, I did not realize I changed the structure of this file. This file is a complete copy and paste. I did not make any manual edits.

**Comment by jlapacik** [comment]
@wolffcm this was mostly just copy paste. There were a few tests that failed due to a bug in flux and one test that failed due to a regression that was reintroduced after the repo split. I was able to find the original fix for the regression and port it (although a few more changes were required to get it working).

I fixed the bug in flux around calculating window boundaries so that the failing launcher tests succeed after updating the flux dependency. I also added an flux end-to-end test to catch any regressions in both repos.

Aside from those fixes, the only other area of manual intervention was from me removing capabilities and unused feature flags and updating the planner tests that used them.

File: tsdb/engine/tsm1/array_cursor.gen.go.tmpl (hunk 2 above)
Comment (jlapacik) [reply tsdb/engine/tsm1/array_cursor.gen.go.tmpl:196]: Correction, this part of the file was not part of the original fix. These two lines do not appear to be in cloud, and so it's not clear to me whether this regression exists there as well. The launcher test that caught it now only lives in OSS.

**Comment by wolffcm** [comment]
@jlapacik Ok, sounds good. The only remaining concern I have is regarding how we upgrade Flux in OSS now.

File: tsdb/engine/tsm1/array_cursor.gen.go [hunk 9]
```diff
@@ -190,6 +190,8 @@ func (c *floatArrayDescendingCursor) reset(seek, end int64, cacheValues Values,
 		})
 		if c.cache.pos == len(c.cache.values) {
 			c.cache.pos--
+		} else if c.cache.values[c.cache.pos].UnixNano() == seek {
+			c.cache.pos--
```
Comment (scbrickley) [inline tsdb/engine/tsm1/array_cursor.gen.go:194]: Another area of code I'm unfamiliar with. What exactly is the purpose of the array cursor type? And why is this `else if` now necessary?

**Review by scbrickley — COMMENTED** [review]
Left a few comments and questions, but really only out of curiosity, rather than critique. Happy to take the conversation offline if it's easier to give me an ELI5 that way.

File: storage/flux/table.go [hunk 10]
```diff
@@ -234,15 +240,27 @@ func (t *floatTable) toArrowBuffer(vs []float64) *array.Float64 {
 func (t *floatGroupTable) toArrowBuffer(vs []float64) *array.Float64 {
 	return arrow.NewFloat(vs, t.alloc)
 }
+func (t *floatWindowSelectorTable) toArrowBuffer(vs []float64) *array.Float64 {
```
[the remark this answers is not in the document]
Comment (jlapacik) [reply storage/flux/table.go:243]: These are used by selector tables to convert a slice of values into an arrow array. These methods are only used by selector table implementations which is why they're just being added right now. Github hides the diffs of generated files by default which is why you might not have noticed these functions being called.

File: tsdb/engine/tsm1/array_cursor.gen.go (hunk 9 above)
Comment (jlapacik) [reply tsdb/engine/tsm1/array_cursor.gen.go:194]: `last()` pushdowns use descending array cursors. If points fall on the boundary of the passed in range, we want to keep the points on the left hand boundary and exclude the points on the right hand boundary. This line concerns the right hand boundary. `seek` represents the `_stop` time. So if we see a point with timestamp equal to `_stop` we want to exclude it.

**Review by wolffcm — APPROVED** [review]
Looks great, nice work.
